## Removing unanswered questions:

In [1]:
import pandas as pd
path = '../Dataset/pythonquestions/Answers.csv'
answers_df = pd.read_csv(path)
answers_df.head()

Id  OwnerUserId          CreationDate  ParentId  Score  \
0  497         50.0  2008-08-02T16:56:53Z       469      4   
1  518        153.0  2008-08-02T17:42:28Z       469      2   
2  536        161.0  2008-08-02T18:49:07Z       502      9   
3  538        156.0  2008-08-02T18:56:56Z       535     23   
4  541        157.0  2008-08-02T19:06:40Z       535     20   

                                                Body  
0  <p>open up a terminal (Applications-&gt;Utilit...  
1  <p>I haven't been able to find anything that d...  
2  <p>You can use ImageMagick's convert utility f...  
3  <p>One possibility is Hudson.  It's written in...  
4  <p>We run <a href="http://buildbot.net/trac">B...

In [2]:
path = '../Dataset/pythonquestions/Questions.csv'
questions_df = pd.read_csv(path)
questions_df.head()

Id  OwnerUserId          CreationDate  Score  \
0  469        147.0  2008-08-02T15:11:16Z     21   
1  502        147.0  2008-08-02T17:01:58Z     27   
2  535        154.0  2008-08-02T18:43:54Z     40   
3  594        116.0  2008-08-03T01:15:08Z     25   
4  683        199.0  2008-08-03T13:19:16Z     28   

                                               Title  \
0  How can I find the full path to a font from it...   
1            Get a preview JPEG of a PDF on Windows?   
2  Continuous Integration System for a Python Cod...   
3     cx_Oracle: How do I iterate over a result set?   
4  Using 'in' to match an attribute of Python obj...   

                                                Body  
0  <p>I am using the Photoshop's javascript API t...  
1  <p>I have a cross-platform (Python) applicatio...  
2  <p>I'm starting work on a hobby project with a...  
3  <p>There are several ways to iterate over a re...  
4  <p>I don't remember whether I was dreaming or ...

In [3]:
answers_df.nunique()

Id              987122
OwnerUserId     149176
CreationDate    984054
ParentId        539238
Score              637
Body            986883
dtype: int64

In [4]:
questions_df.nunique()

Id              607282
OwnerUserId     213927
CreationDate    606068
Score              503
Title           606928
Body            607243
dtype: int64

So, there are a total of 607282 questions and 987122 answers. Finding out unanswered questions:

In [5]:
answers_df.sort_values('ParentId', inplace=True)
answers_df.head()

Id  OwnerUserId          CreationDate  ParentId  Score  \
0        497         50.0  2008-08-02T16:56:53Z       469      4   
1        518        153.0  2008-08-02T17:42:28Z       469      2   
3003  195170        745.0  2008-10-12T07:02:40Z       469      1   
38      3040        457.0  2008-08-06T03:01:23Z       469     12   
2        536        161.0  2008-08-02T18:49:07Z       502      9   

                                                   Body  
0     <p>open up a terminal (Applications-&gt;Utilit...  
1     <p>I haven't been able to find anything that d...  
3003  <p>There must be a method in Cocoa to get a li...  
38    <p>Unfortunately the only API that isn't depre...  
2     <p>You can use ImageMagick's convert utility f...

In [6]:
qIDs_withAnswers = set(answers_df['ParentId'].tolist())
print "qIDs with answers: ", len(qIDs_withAnswers)
total_qIDs = set(questions_df['Id'].tolist())
print "Total qIDs: ", len(total_qIDs)

qIDs with answers:  539238
Total qIDs:  607282


In [7]:
607282 - 539238

68044

Hence, there are 68,044 unanswered questions. Removing them now.

In [8]:
unanswered_qIDs = list(total_qIDs - qIDs_withAnswers)
print len(unanswered_qIDs)

68044


In [9]:
# REMOVING 'unanswered_qIDs' from the main 'questions_df'
clean_questions_df = questions_df[~questions_df['Id'].isin(unanswered_qIDs)]
clean_questions_df

Id  OwnerUserId          CreationDate  Score  \
0            469        147.0  2008-08-02T15:11:16Z     21   
1            502        147.0  2008-08-02T17:01:58Z     27   
2            535        154.0  2008-08-02T18:43:54Z     40   
3            594        116.0  2008-08-03T01:15:08Z     25   
4            683        199.0  2008-08-03T13:19:16Z     28   
5            742        189.0  2008-08-03T15:55:28Z     30   
6            766    1384652.0  2008-08-03T17:44:07Z     20   
7            773        207.0  2008-08-03T18:27:09Z    256   
8            972        145.0  2008-08-04T02:17:51Z    364   
9           1476         92.0  2008-08-04T18:20:36Z    251   
10          1734         59.0  2008-08-05T00:12:55Z     20   
11          1829         30.0  2008-08-05T02:39:23Z     12   
12          1854        116.0  2008-08-05T03:23:18Z    270   
13          1983        116.0  2008-08-05T07:18:55Z     31   
14          2311        394.0  2008-08-05T13:40:47Z     26   
15          2933    1384652.0  2008-08-05T22:26:00Z    171   
16          3061        121.0  2008-08-06T03:36:08Z    761   
17          3976        556.0  2008-08-06T21:00:01Z     12   
18          4942    1384652.0  2008-08-07T16:43:21Z      5   
19          5102        556.0  2008-08-07T18:24:12Z     14   
20          5136        676.0  2008-08-07T18:47:58Z      6   
21          5313        680.0  2008-08-07T21:07:24Z     11   
22          5415        680.0  2008-08-07T22:24:27Z     21   
23          5419        207.0  2008-08-07T22:26:58Z     72   
24          5909        394.0  2008-08-08T13:35:19Z     28   
25          5966        745.0  2008-08-08T14:05:45Z     10   
26          8154        736.0  2008-08-11T19:59:32Z     10   
27          8692          NaN  2008-08-12T11:28:36Z    164   
28          8948        145.0  2008-08-12T15:16:00Z     81   
29         10123        742.0  2008-08-13T17:46:41Z      8   
...          ...          ...                   ...    ...   
607213  40141705    2695332.0  2016-10-19T21:19:34Z      0   
607217  40141729    1107049.0  2016-10-19T21:22:00Z      0   
607219  40141753    4621508.0  2016-10-19T21:23:57Z      0   
607221  40141828    1230911.0  2016-10-19T21:30:25Z      0   
607222  40141856     512652.0  2016-10-19T21:31:52Z      2   
607225  40141881    2336654.0  2016-10-19T21:34:18Z      4   
607226  40141895    2201603.0  2016-10-19T21:35:05Z      1   
607229  40142004    3965778.0  2016-10-19T21:43:45Z      1   
607230  40142024    1290147.0  2016-10-19T21:44:57Z      1   
607234  40142166    7044761.0  2016-10-19T21:56:19Z     -1   
607235  40142194    7044759.0  2016-10-19T21:58:32Z      1   
607236  40142236    7044735.0  2016-10-19T22:01:05Z     -3   
607238  40142259    6273496.0  2016-10-19T22:02:40Z      0   
607241  40142302    1107049.0  2016-10-19T22:05:58Z      1   
607246  40142380    4495740.0  2016-10-19T22:13:40Z      0   
607248  40142476    6661072.0  2016-10-19T22:21:57Z      1   
607249  40142494    6930377.0  2016-10-19T22:23:40Z      1   
607250  40142538    5992039.0  2016-10-19T22:28:03Z      0   
607254  40142646    4808466.0  2016-10-19T22:38:43Z      0   
607256  40142686    4150912.0  2016-10-19T22:42:47Z      1   
607258  40142731    6875348.0  2016-10-19T22:46:59Z      0   
607260  40142752    6934489.0  2016-10-19T22:49:06Z      0   
607262  40142804    6902762.0  2016-10-19T22:54:26Z      0   
607263  40142811    2679662.0  2016-10-19T22:55:05Z      2   
607265  40142901    7044936.0  2016-10-19T23:05:07Z     -2   
607266  40142906    7044946.0  2016-10-19T23:05:34Z      2   
607268  40142948    4307777.0  2016-10-19T23:09:22Z     -2   
607272  40143133    6273496.0  2016-10-19T23:29:37Z      1   
607275  40143166    7044992.0  2016-10-19T23:33:31Z      1   
607277  40143190     333403.0  2016-10-19T23:36:01Z      1   

                                                    Title  \
0       How can I find the full path to a font from it...   
1                 Get a preview JPEG of a PDF o

In [10]:
# Another way to do the same thing:
blah_df = questions_df[questions_df['Id'].isin(qIDs_withAnswers)]
# Check if both are equal:
print blah_df.equals(clean_questions_df)

True


In [11]:
# Final check: everything in clean_questions_df should be in answers_df, and vice-versa.
print set(clean_questions_df['Id'].tolist()) - set(answers_df['ParentId'])
print set(answers_df['ParentId'].tolist()) - set(clean_questions_df['Id'])

set([])
set([])


In [12]:
# Writing it to a CSV
clean_questions_df.to_csv('../Dataset/pythonquestions/cleanQuestions.csv', index=False)

## Now, labelling the dataset with "BestAnswer/NonBestAnswer"

In [13]:
answers_df.head(10)

Id  OwnerUserId          CreationDate  ParentId  Score  \
0        497         50.0  2008-08-02T16:56:53Z       469      4   
1        518        153.0  2008-08-02T17:42:28Z       469      2   
3003  195170        745.0  2008-10-12T07:02:40Z       469      1   
38      3040        457.0  2008-08-06T03:01:23Z       469     12   
2        536        161.0  2008-08-02T18:49:07Z       502      9   
70      7090         13.0  2008-08-10T08:08:33Z       502     25   
69      7073        878.0  2008-08-10T07:10:19Z       502      2   
3        538        156.0  2008-08-02T18:56:56Z       535     23   
4        541        157.0  2008-08-02T19:06:40Z       535     20   
709    61746       6372.0  2008-09-15T00:11:21Z       535     13   

                                                   Body  
0     <p>open up a terminal (Applications-&gt;Utilit...  
1     <p>I haven't been able to find anything that d...  
3003  <p>There must be a method in Cocoa to get a li...  
38    <p>Unfortunately the only API that isn't depre...  
2     <p>You can use ImageMagick's convert utility f...  
70    <p>ImageMagick delegates the PDF->bitmap conve...  
69    <p>Is the PC likely to have Acrobat installed?...  
3     <p>One possibility is Hudson.  It's written in...  
4     <p>We run <a href="http://buildbot.net/trac">B...  
709   <p>We use both Buildbot and Hudson for Jython ...

In [14]:
# answers_df.sort_values(['ParentId','Score'], ascending=[True, False], inplace=True)

In [15]:
idx = answers_df.groupby(['ParentId'])['Score'].transform(max) == answers_df['Score']
idx

0         False
1         False
3003      False
38         True
2         False
70         True
69        False
3          True
4         False
709       False
6         False
926       False
1576      False
197542    False
1773      False
203       False
5          True
112050    False
10        False
8         False
7         False
642        True
9         False
245       False
244       False
77        False
273        True
11        False
1746      False
42        False
          ...  
987082     True
987077    False
987080     True
987075     True
987086     True
987107     True
987110    False
987085     True
987084    False
987089    False
987087     True
987121     True
987094    False
987102    False
987099     True
987103     True
987101     True
987115     True
987095     True
987096     True
987098    False
987100     True
987109     True
987117    False
987104     True
987111     True
987114     True
987118     True
987113     True
987116    False
Name: Score, Length: 987

In [16]:
bestAnswer_df = answers_df[idx]
bestAnswer_df.head(10)

Id  OwnerUserId          CreationDate  ParentId  Score  \
38    3040        457.0  2008-08-06T03:01:23Z       469     12   
70    7090         13.0  2008-08-10T08:08:33Z       502     25   
3      538        156.0  2008-08-02T18:56:56Z       535     23   
5      595        116.0  2008-08-03T01:17:36Z       594     25   
642  57833       4702.0  2008-09-11T22:42:14Z       683     29   
273  33957       3207.0  2008-08-29T04:29:22Z       742     36   
20    1619         92.0  2008-08-04T21:54:11Z       766     17   
71    7286        207.0  2008-08-10T18:45:32Z       773    394   
37    2982         99.0  2008-08-06T00:33:35Z       972    576   
103  13107       1531.0  2008-08-16T12:35:36Z      1476    204   

                                                  Body  
38   <p>Unfortunately the only API that isn't depre...  
70   <p>ImageMagick delegates the PDF->bitmap conve...  
3    <p>One possibility is Hudson.  It's written in...  
5    <p>The canonical way is to use the built-in cu...  
642  <p>Using a list comprehension would build a te...  
273  <p>I've created and used my own generic view c...  
20   <p>MySQLdb is what I have used before.</p>\r\n...  
71   <p>As Sebastjan said, <strong>you first have t...  
37   <p>In Python, there is a difference between fu...  
103  <p>For reference&mdash;<em>future</em> Python ...

In [17]:
bestAnswer_df.nunique()

Id              614567
OwnerUserId     100735
CreationDate    613401
ParentId        539238
Score              589
Body            614517
dtype: int64

In [18]:
clean_questions_df.nunique()

Id              539238
OwnerUserId     192929
CreationDate    538339
Score              503
Title           538918
Body            539215
dtype: int64

Currently, we have 614567 best answers and 539238 questions which is not possible. This is due to answers (for the same question) having the same number of upvotes. Removing duplicates:

In [19]:
len(set(bestAnswer_df['ParentId'].tolist())) # set() should have the same number: 539238

539238

In [20]:
# Remove duplicates:
bestAnswer_df.drop_duplicates(subset='ParentId', keep="last", inplace=True)
bestAnswer_df.head()

/Users/sunyambagga/anaconda/lib/python2.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


Id  OwnerUserId          CreationDate  ParentId  Score  \
38    3040        457.0  2008-08-06T03:01:23Z       469     12   
70    7090         13.0  2008-08-10T08:08:33Z       502     25   
3      538        156.0  2008-08-02T18:56:56Z       535     23   
5      595        116.0  2008-08-03T01:17:36Z       594     25   
642  57833       4702.0  2008-09-11T22:42:14Z       683     29   

                                                  Body  
38   <p>Unfortunately the only API that isn't depre...  
70   <p>ImageMagick delegates the PDF->bitmap conve...  
3    <p>One possibility is Hudson.  It's written in...  
5    <p>The canonical way is to use the built-in cu...  
642  <p>Using a list comprehension would build a te...

In [21]:
bestAnswer_ids = set(bestAnswer_df['Id'].tolist())
bestAnswer_df.nunique()

Id              539238
OwnerUserId      91333
CreationDate    538410
ParentId        539238
Score              589
Body            539200
dtype: int64

In [22]:
nonBestAnswer_ids = set(answers_df['Id'].tolist()) - bestAnswer_ids
nonBestAnswer_ids

{12080992,
 11534342,
 9437192,
 9437193,
 33554449,
 9437202,
 38972078,
 12582937,
 16777244,
 5941936,
 22020130,
 2097190,
 5941937,
 2097197,
 27263024,
 27263025,
 37748786,
 30408755,
 19922997,
 31457335,
 29360185,
 3145790,
 13631556,
 35651653,
 36700230,
 28311625,
 30408779,
 3145804,
 36700237,
 19923023,
 2097234,
 39845975,
 4194392,
 9437273,
 19923034,
 11534427,
 30583482,
 14680161,
 1048674,
 27263075,
 14680165,
 35651686,
 1048685,
 12583023,
 33903976,
 26214516,
 1048693,
 30408825,
 1048702,
 36700288,
 28311683,
 28311685,
 3145862,
 1048714,
 39846029,
 10485907,
 524313,
 14680216,
 31457435,
 1048732,
 14680225,
 5243043,
 1048740,
 24117413,
 15728807,
 21146311,
 17825964,
 17825966,
 35651761,
 1048754,
 13631667,
 7514825,
 16777400,
 3145913,
 8388795,
 3145917,
 3145918,
 17825993,
 7340234,
 30408910,
 24117456,
 32960864,
 36700370,
 7340245,
 3145944,
 13631706,
 29360347,
 28311773,
 22020321,
 1048802,
 23068899,
 27263204,
 32506088,
 29705080,

In [23]:
len(nonBestAnswer_ids)

447884

In [24]:
447884 + 539238

987122

In [25]:
answers_df.nunique()

Id              987122
OwnerUserId     149176
CreationDate    984054
ParentId        539238
Score              637
Body            986883
dtype: int64

In [26]:
# Writing best answers to a 2-column CSV:
with open('../Dataset/pythonquestions/labeled_answerIDs.csv', 'wb') as f:
    f.write("Answer_ID,Label\n")
    
    for ID in bestAnswer_ids:
        f.write(str(ID)+","+"BestAnswer"+"\n")
        
    for ID in nonBestAnswer_ids:
        f.write(str(ID)+","+"Non_BestAnswer"+"\n")

In [27]:
temp_df = pd.read_csv('../Dataset/pythonquestions/labeled_answerIDs.csv')
temp_df

Answer_ID           Label
0         5242883      BestAnswer
1        31457284      BestAnswer
2        33344935      BestAnswer
3         7716472      BestAnswer
4        20971530      BestAnswer
5        35651595      BestAnswer
6         7340044      BestAnswer
7        15728656      BestAnswer
8        19922963      BestAnswer
9        14854830      BestAnswer
10       35651608      BestAnswer
11       32505860      BestAnswer
12       39001354      BestAnswer
13        3145756      BestAnswer
14       36382642      BestAnswer
15       12582945      BestAnswer
16       28311587      BestAnswer
17       23068709      BestAnswer
18        7340071      BestAnswer
19       26214440      BestAnswer
20       16777257      BestAnswer
21       34603051      BestAnswer
22       10485804      BestAnswer
23       34777778      BestAnswer
24        7689565      BestAnswer
25       37748785      BestAnswer
26       26214450      BestAnswer
27       34603061      BestAnswer
28       30408761      BestAnswer
29       33554490      BestAnswer
...           ...             ...
987092   37748675  Non_BestAnswer
987093   25837958  Non_BestAnswer
987094   38797255  Non_BestAnswer
987095   14233131  Non_BestAnswer
987096    4019532  Non_BestAnswer
987097    3145675  Non_BestAnswer
987098   35651532  Non_BestAnswer
987099   32331085  Non_BestAnswer
987100   27234429  Non_BestAnswer
987101   19922897  Non_BestAnswer
987102   25165778  Non_BestAnswer
987103   26214356  Non_BestAnswer
987104   35651543  Non_BestAnswer
987105   35651545  Non_BestAnswer
987106   17825754  Non_BestAnswer
987107   11534299  Non_BestAnswer
987108    7339997  Non_BestAnswer
987109     699045  Non_BestAnswer
987110   12297392  Non_BestAnswer
987111   11534308  Non_BestAnswer
987112    5242855  Non_BestAnswer
987113    1048557  Non_BestAnswer
987114   24117231  Non_BestAnswer
987115    1048564  Non_BestAnswer
987116   31457269  Non_BestAnswer
987117    5242873  Non_BestAnswer
987118   16696800  Non_BestAnswer
987119    3145724  Non_BestAnswer
987120    2097149  Non_BestAnswer
987121   12582911  Non_BestAnswer

[987122 rows x 2 columns]

In [28]:
ba = temp_df[temp_df['Label'] == 'BestAnswer']
ba.nunique()

Answer_ID    539238
Label             1
dtype: int64

In [29]:
nba = temp_df[temp_df['Label'] == 'Non_BestAnswer']
nba.nunique()

Answer_ID    447884
Label             1
dtype: int64

In [30]:
# Code to retrieve all columns given only answer IDs:
#answers_df[answers_df['Id'].isin(list(nonBestAnswer_ids))]